toolchain/check/pattern_match.cpp

```diff
@@ -290,39 +295,38 @@ auto CalleePatternMatch(Context& context,
                         SemIR::InstBlockId param_patterns_id,
                         SemIR::InstId return_slot_pattern_id)
     -> ParameterBlocks {
-  auto params_id = SemIR::InstBlockId::Invalid;
-  auto implicit_params_id = SemIR::InstBlockId::Invalid;
-
   MatchContext match(MatchKind::Callee);
 
-  if (implicit_param_patterns_id.is_valid()) {
-    // We add work to the stack in reverse so that the results will be produced
-    // in the original order.
+  // We add work to the stack in reverse so that the results will be produced
+  // in the original order.
+  if (return_slot_pattern_id.is_valid()) {
+    match.AddWork({.pattern_id = return_slot_pattern_id,
+                   .scrutinee_id = SemIR::InstId::Invalid});
+  }
+
+  if (param_patterns_id.is_valid()) {
     for (SemIR::InstId inst_id :
-         llvm::reverse(context.inst_blocks().Get(implicit_param_patterns_id))) {
+         llvm::reverse(context.inst_blocks().Get(param_patterns_id))) {
       match.AddWork(
           {.pattern_id = inst_id, .scrutinee_id = SemIR::InstId::Invalid});
     }
-    implicit_params_id = match.DoWork(context);
   }
 
-  if (param_patterns_id.is_valid()) {
+  if (implicit_param_patterns_id.is_valid()) {
     for (SemIR::InstId inst_id :
-         llvm::reverse(context.inst_blocks().Get(param_patterns_id))) {
+         llvm::reverse(context.inst_blocks().Get(implicit_param_patterns_id))) {
       match.AddWork(
           {.pattern_id = inst_id, .scrutinee_id = SemIR::InstId::Invalid});
     }
-    params_id = match.DoWork(context);
   }
 
-  if (return_slot_pattern_id.is_valid()) {
-    match.AddWork({.pattern_id = return_slot_pattern_id,
-                   .scrutinee_id = SemIR::InstId::Invalid});
-    CARBON_CHECK(match.DoWork(context) == SemIR::InstBlockId::Empty);
+  if (!return_slot_pattern_id.is_valid() && !param_patterns_id.is_valid() &&
+      !implicit_param_patterns_id.is_valid()) {
+    return {.calling_convention_params_id = SemIR::InstBlockId::Invalid,
+            .return_slot_id = SemIR::InstId::Invalid};
```

Comment:
Had you considered checking this at the top, before the `MatchContext` construction?

Reply:
Yeah, that's a bit cleaner. Done.

Reply:
Why do you still construct `MatchContext` before the comparison?

Reply:
No good reason. Fixed.